subsys/pcd/src/pcd.c

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <zephyr.h>
+#include <device.h>
+#include <dfu/pcd.h>
+#include <logging/log.h>
+#include <storage/stream_flash.h>
+
+LOG_MODULE_REGISTER(pcd, CONFIG_PCD_LOG_LEVEL);
+
+/** Magic value written to indicate that a copy should take place. */
+#define PCD_CMD_MAGIC_COPY 0xb5b4b3b6
+/** Magic value written to indicate that a something failed. */
+#define PCD_CMD_MAGIC_FAIL 0x25bafc15
+/** Magic value written to indicate that a copy is done. */
+#define PCD_CMD_MAGIC_DONE 0xf103ce5d
+
+#if defined(CONFIG_SOC_NRF5340_CPUAPP) && defined(CONFIG_MCUBOOT)
+#include <pm_config.h>
+#include <hal/nrf_reset.h>
+#include <hal/nrf_spu.h>
+/** Offset which the application should be copied into */
+#define NET_CORE_APP_OFFSET PM_CPUNET_B0N_SIZE
+#endif
+
+struct pcd_cmd {
+	uint32_t magic; /* Magic value to identify this structure in memory */
+	const void *data;     /* Data to copy*/
+	size_t len;           /* Number of bytes to copy */
+	off_t offset;         /* Offset to store the flash image in */
+} __aligned(4);
+
+static struct pcd_cmd *cmd = (struct pcd_cmd *)PCD_CMD_ADDRESS;
+
+void pcd_fw_copy_invalidate(void)
+{
+	cmd->magic = PCD_CMD_MAGIC_FAIL;
+}
+
+enum pcd_status pcd_fw_copy_status_get(void)
+{
+	if (cmd->magic == PCD_CMD_MAGIC_COPY) {
+		return PCD_STATUS_COPY;
+	} else if (cmd->magic == PCD_CMD_MAGIC_DONE) {
+		return PCD_STATUS_COPY_DONE;
+	}
+
+	return PCD_STATUS_COPY_FAILED;
+}
+
+int pcd_fw_copy(struct device *fdev)
+{
+	struct stream_flash_ctx stream;
+	uint8_t buf[CONFIG_PCD_BUF_SIZE];
+	int rc;
+
```

Comment:
Maybe check that `cmd->magic` is `PCD_CMD_MAGIC_COPY` here?

Reply:
Added error value return if not, IMO it should be like that to give the caller feedback about the failure.